WordPress/Classes/Services/AccountService.m

```diff
@@ -180,27 +177,69 @@ - (WPAccount *)createOrUpdateAccountWithUsername:(NSString *)username
 
     if (!account) {
         account = [NSEntityDescription insertNewObjectForEntityForName:@"Account" inManagedObjectContext:self.managedObjectContext];
-        account.uuid = [[NSUUID new] UUIDString];
         account.username = username;
     }
     account.authToken = authToken;
     [[ContextManager sharedInstance] saveContextAndWait:self.managedObjectContext];
 
-    if (![self defaultWordPressComAccount]) {
-        [self setDefaultWordPressComAccount:account];
-    }
+    [self setDefaultWordPressComAccount:account];
 
     return account;
 }
 
+- (void)retrieveAllAccountsWith:(WPAccountRetrieveCompletion)completion
+{
+    NSManagedObjectContext *context = [[ContextManager sharedInstance] mainContext];
+    NSFetchRequest *request = [NSFetchRequest fetchRequestWithEntityName:@"Account"];
+    [request setIncludesSubentities:YES];
+    [request setReturnsObjectsAsFaults:NO];
+    NSError *error;
+    NSArray* accounts = [context executeFetchRequest:request error:&error];
+
+    NSMutableArray *mutableCompletedAccounts = [NSMutableArray array];
+    __block NSInteger callbackCounter = 0;
+    for(WPAccount* account in accounts) {
+
+        if (account.email == nil) {
+
+            [self updateUserDetailsForAccount:account
+                                      success:^{
+                                          [mutableCompletedAccounts addObject:account];
+                                          callbackCounter++;
+                                          if (callbackCounter == accounts.count) {
+                                              completion(accounts);
+                                          }
+                                      } failure:^(NSError * _Nonnull error) {
+                                          callbackCounter++;
+                                          if (callbackCounter == accounts.count) {
+                                              completion(mutableCompletedAccounts);
+                                          }
+                                      }];
+        }
+        else {
+            [mutableCompletedAccounts addObject:account];
+            callbackCounter++;
+            if (callbackCounter == accounts.count) {
+                completion(mutableCompletedAccounts);
+            }
+        }
+    }
+
+    if (accounts.count == 0) {
+        completion(@[]);
+    }
+}
+
 - (NSUInteger)numberOfAccounts
 {
+    NSManagedObjectContext *context = [[ContextManager sharedInstance] mainContext];
```

Comment:
When a service is created it can be passed a reference to the main context, or to a derived content (a child of the main context).  Using the main context, here when a derived context was passed could introduce some unexpected issues.  It would be better to always use `self.managedObjectContext` for this reason.